File: packages/julia/src/Seguranca/Ataques/AtaqueCorrelacaoPosicoes.jl
"""
Verifica se posições DIFERENTES dentro do mesmo bloco de 32 bytes saem
correlacionadas (ex: posição 0 acompanha posição 16). Uma correlação entre
posições de saída é exatamente o tipo de estrutura que o antigo bug da
"metade igual" produzia - e que a autocorrelação temporal pode não pegar.
"""
struct AtaqueCorrelacaoPosicoes <: Ataque
    amostras::Int
    tamanho_bloco::Int
    limite_correlacao::Float64
end

AtaqueCorrelacaoPosicoes(;
    amostras::Integer = 3000,
    tamanho_bloco::Integer = 32,
    limite_correlacao::Real = 0.15,
) = AtaqueCorrelacaoPosicoes(
    Int(amostras),
    Int(tamanho_bloco),
    Float64(limite_correlacao),
)

nome(::AtaqueCorrelacaoPosicoes) = "Correlação entre posições do bloco"

function _pearson_posicoes(blocos::AbstractVector, a::Integer, b::Integer)
    n = length(blocos)
    ma = 0.0
    mb = 0.0
    for d in blocos
        ma += d[a+1]
        mb += d[b+1]
    end
    ma /= n
    mb /= n

    num = 0.0
    da = 0.0
    db = 0.0
    for d in blocos
        xa = d[a+1] - ma
        xb = d[b+1] - mb
        num += xa * xb
        da += xa * xa
        db += xb * xb
    end

    return da > 0 && db > 0 ? num / sqrt(da * db) : 0.0
end

function executar(a::AtaqueCorrelacaoPosicoes, alvo::AlvoCriptografico)
    chave = chave_de_teste(alvo)

    blocos = Vector{UInt8}[]
    for _ in 1:(a.amostras)
        push!(
            blocos,
            gerar_keystream_bruto(
                alvo,
                chave,
                random_bytes(tamanho_iv(alvo)),
                "enc",
                a.tamanho_bloco,
            ),
        )
    end

    suspeitos = String[]
    pior = 0.0

    for a_pos in 0:(a.tamanho_bloco-1)
        for b_pos in (a_pos+1):(a.tamanho_bloco-1)
            corr = _pearson_posicoes(blocos, a_pos, b_pos)
            if abs(corr) > abs(pior)
                pior = corr
            end
            if abs(corr) > a.limite_correlacao
                push!(suspeitos, "posições $a_pos-$b_pos (r=$(fmt(corr, 3)))")
            end
        end
    end

    if !isempty(suspeitos)
        return ResultadoAtaque(
            nome(a),
            true,
            "alta",
            "$(length(suspeitos)) par(es) correlacionado(s): $(join(suspeitos[1:min(10, length(suspeitos))], "; "))",
            Dict{String,Any}("suspeitos" => suspeitos),
        )
    end

    return ResultadoAtaque(
        nome(a),
        false,
        "info",
        "Nenhum par de posições correlacionado acima de $(fmt(a.limite_correlacao, 2)) (pior |r|=$(fmt(abs(pior), 3)))",
    )
end

File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueCorrelacaoPosicoes.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Verifica se posições DIFERENTES dentro do mesmo bloco de 32 bytes saem
 * correlacionadas (ex: posição 0 acompanha posição 16). Uma correlação entre
 * posições de saída é exatamente o tipo de estrutura que o antigo bug da
 * "metade igual" produzia - e que a autocorrelação temporal pode não pegar.
 */
public class AtaqueCorrelacaoPosicoes implements AtaqueInterface {
    private final int amostras;
    private final int tamanhoBloco;
    private final double limiteCorrelacao;

    public AtaqueCorrelacaoPosicoes() {
        this(3000, 32, 0.15);
    }

    public AtaqueCorrelacaoPosicoes(int amostras, int tamanhoBloco, double limiteCorrelacao) {
        this.amostras = amostras;
        this.tamanhoBloco = tamanhoBloco;
        this.limiteCorrelacao = limiteCorrelacao;
    }

    @Override
    public String nome() {
        return "Correlação entre posições do bloco";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] chave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8);
        byte[] primeiro = alvo.gerarKeystreamBruto(
                chave, Util.randomBytes(alvo.tamanhoIv()), "enc", this.tamanhoBloco);
        if (primeiro == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        List<byte[]> blocos = new ArrayList<>();
        for (int i = 0; i < this.amostras; i++) {
            blocos.add(alvo.gerarKeystreamBruto(
                    chave, Util.randomBytes(alvo.tamanhoIv()), "enc", this.tamanhoBloco));
        }

        List<String> suspeitos = new ArrayList<>();
        double pior = 0.0;

        for (int a = 0; a < this.tamanhoBloco; a++) {
            for (int b = a + 1; b < this.tamanhoBloco; b++) {
                double corr = this.pearson(blocos, a, b);
                if (Math.abs(corr) > Math.abs(pior)) {
                    pior = corr;
                }
                if (Math.abs(corr) > this.limiteCorrelacao) {
                    suspeitos.add(String.format(Locale.ROOT, "posições %d-%d (r=%.3f)", a, b, corr));
                }
            }
        }

        if (!suspeitos.isEmpty()) {
            return new ResultadoAtaque(
                    nome(),
                    true,
                    Severidade.ALTA,
                    suspeitos.size() + " par(es) correlacionado(s): "
                            + String.join("; ", suspeitos.subList(0, Math.min(10, suspeitos.size()))),
                    Map.of("suspeitos", suspeitos));
        }

        return new ResultadoAtaque(
                nome(),
                false,
                Severidade.INFO,
                String.format(
                        Locale.ROOT,
                        "Nenhum par de posições correlacionado acima de %.2f (pior |r|=%.3f)",
                        this.limiteCorrelacao,
                        Math.abs(pior)));
    }

    private double pearson(List<byte[]> blocos, int a, int b) {
        int n = blocos.size();
        double ma = 0.0;
        double mb = 0.0;
        for (byte[] d : blocos) {
            ma += (d[a] & 0xFF);
            mb += (d[b] & 0xFF);
        }
        ma /= n;
        mb /= n;

        double num = 0.0;
        double da = 0.0;
        double db = 0.0;
        for (byte[] d : blocos) {
            double xa = (d[a] & 0xFF) - ma;
            double xb = (d[b] & 0xFF) - mb;
            num += xa * xb;
            da += xa * xa;
            db += xb * xb;
        }

        return da > 0 && db > 0 ? num / Math.sqrt(da * db) : 0.0;
    }
}
